### PR TITLE
Add RepositoryServer ObjectReference in Actionset

### PR DESCRIPTION
**Pull request by r4rajat**
## Change Overview

Add RepositoryServer ObjectReference in ```kanctl``` command line to create various actions.
This PR adds support for passing `--repository-server` flag in `kanctl` command line

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [x] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues 

- fixes #issue-number

## Test Plan

- [x] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

#### Build ```kanctl``` command line tool

```
go build -o kanctl cmd/kanctl/main.go
```

#### Run Actionset

```
./kanctl create actionset --action backup --namespace kanister --blueprint backupdate-bp --deployment time-logger/time-logger --repository-server=kopia-repo-server-1

actionset backup-czvrz created
```

Also there's a check which ensures that Repository Server CR is in ```ServerReady``` state before creating actionset. 

```
./kanctl create actionset --action backup --namespace kanister --blueprint backupdate-bp --deployment time-logger/time-logger --repository-server=kopia-repo-server-1

Please make sure that Repository Server CR 'kopia-repo-server-1' is in Ready State: Repository Server Not Ready
Error: resource verification failed
```

### CODE REVIEW

File: pkg/customresource/actionset.yaml [hunk 1]
```diff
@@ -11,49 +11,78 @@ spec:
...
-      openAPIV3Schema:
-        properties:
-          apiVersion:
-            description: 'APIVersion defines the versioned schema of this representation
+    - name: v1alpha1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          properties:
```
Comment (viveksinghggits) [inline pkg/customresource/actionset.yaml:19]: Do we need changes in this file @r4rajat ?

Comment (r4rajat) [reply pkg/customresource/actionset.yaml:19]: Yes @viveksinghggits, as we're adding RepositoryServer field in actionset, so we've to modify it's customresource definition

File: pkg/customresource/actionset.yaml [hunk 2]
```diff
@@ -11,49 +11,78 @@ spec:
...
-    served: true
-    storage: true
-    schema:
-      openAPIV3Schema:
-        properties:
-          apiVersion:
-            description: 'APIVersion defines the versioned schema of this representation
+    - name: v1alpha1
+      served: true
+      storage: true
```
Comment (viveksinghggits) [inline pkg/customresource/actionset.yaml:16]: We generate this yaml right?
Can we just update the new entry manually here?
We can generate the yaml but just copy the snipped for `RepositoryServer` and copy that here. We wont see all these unnecessary changes here.
Or if we are sure that this wouldn't result into any problem, which it wont I think, we can go ahead with this as well.

File: pkg/kanctl/actionset.go [hunk 3]
```diff
@@ -92,6 +94,7 @@ func newActionSetCmd() *cobra.Command {
 	cmd.Flags().StringSliceP(deploymentFlagName, "d", []string{}, "deployment for the action set, comma separated namespace/name pairs (eg: --deployment namespace1/name1,namespace2/name2)")
 	cmd.Flags().StringSliceP(optionsFlagName, "o", []string{}, "specify options for the action set, comma separated key=value pairs (eg: --options key1=value1,key2=value2)")
 	cmd.Flags().StringP(profileFlagName, "p", "", "profile for the action set")
+	cmd.Flags().StringP(repositoryServerFlagName, "r", "", "kopia repository server custom resource reference")
```
Comment (viveksinghggits) [inline pkg/kanctl/actionset.go:97]: Let's also specify the format here.
For, like its specified for other flags.
```
eg: --deployment namespace1/name1,namespace2/name2
```

File: pkg/kanctl/actionset.go [hunk 4]
```diff
@@ -342,6 +355,25 @@ func parseProfile(cmd *cobra.Command, ns string) (*crv1alpha1.ObjectReference, e
...
 
+func parseRepositoryServer(cmd *cobra.Command, ns string) (*crv1alpha1.ObjectReference, error) {
+	repositoryServerName, _ := cmd.Flags().GetString(repositoryServerFlagName)
+	if repositoryServerName == "" {
+		return nil, nil
+	}
+	if strings.Contains(repositoryServerName, "/") {
+		temp := strings.Split(repositoryServerName, "/")
+		if len(temp) != 2 {
+			return nil, errors.Errorf("Invalid repository server name %s, should be of the form ( --repository-server namespace/name OR --repository-server name)", repositoryServerName)
```
Comment (viveksinghggits) [inline pkg/kanctl/actionset.go:366]:
```suggestion
			return nil, errors.Errorf("Invalid repository server name %s, it should be of the form ( --repository-server namespace/name or --repository-server name)", repositoryServerName)
```

File: pkg/kanctl/actionset.go [hunk 5]
```diff
@@ -342,6 +355,25 @@ func parseProfile(cmd *cobra.Command, ns string) (*crv1alpha1.ObjectReference, e
 	}, nil
 }
 
+func parseRepositoryServer(cmd *cobra.Command, ns string) (*crv1alpha1.ObjectReference, error) {
```
Comment (viveksinghggits) [inline pkg/kanctl/actionset.go:358]: We don't  have to do anything here, but I would rather prefer `objectRefFromRepoServer` instead of `parseRepositoryServer`.
But I understand is ok to keep it as it is for consistency. Maybe just add a comment that it returns object reference for passed repo server.

File: pkg/kanctl/actionset.go [hunk 6]
```diff
@@ -631,6 +664,21 @@ func verifyParams(ctx context.Context, p *PerformParams, cli kubernetes.Interfac
 		}
 	}()
 
+	// RepositoryServer
+	go func() {
+		defer wg.Done()
+		if p.RepositoryServer != nil {
+			rs, err := crCli.CrV1alpha1().RepositoryServers(p.RepositoryServer.Namespace).Get(ctx, p.RepositoryServer.Name, metav1.GetOptions{})
+			if err != nil {
```
Comment (viveksinghggits) [inline pkg/kanctl/actionset.go:672]: Should we check if the error actually says that the resource is not found. So we would have three categories
- Resource is not found
- There was an error getting the resource
- Resource is found and not ready

You don't have to do it, let's  wait for others opinions.

File: pkg/param/param.go [hunk 7]
```diff
@@ -299,6 +350,14 @@ func fetchSecretCredential(ctx context.Context, cli kubernetes.Interface, sr *cr
 	}, nil
 }
 
+func fetchSecretFromSecretRef(ctx context.Context, cli kubernetes.Interface, ref v1.SecretReference) (*v1.Secret, error) {
```
Comment (viveksinghggits) [inline pkg/param/param.go:353]:
```suggestion
func secretFromSecretRef(ctx context.Context, cli kubernetes.Interface, ref v1.SecretReference) (*v1.Secret, error) {
```

File: pkg/kanctl/actionset.go (hunk 3 above)
Comment (r4rajat) [reply pkg/kanctl/actionset.go:97]: Sure wil add that

File: pkg/kanctl/actionset.go (hunk 6 above)
Comment (r4rajat) [reply pkg/kanctl/actionset.go:672]: - So the checks for first scenerio, ```Resource is not found``` and the third scenerio ```Resource is found and not ready``` has been handled while creating actionset, we make sure Resources are Present and Resources are in ready state
- And the second scenerio, ```There was an error getting the resource``` is handled in the code above

File: pkg/customresource/actionset.yaml (hunk 2 above)
Comment (r4rajat) [reply pkg/customresource/actionset.yaml:16]: Sure will try that !

File: pkg/kanctl/actionset.go (hunk 5 above)
Comment (r4rajat) [reply pkg/kanctl/actionset.go:358]: Added the comment !

File: pkg/kanctl/actionset.go (hunk 4 above)
Comment (r4rajat) [reply pkg/kanctl/actionset.go:366]: Will make changes

File: pkg/kanctl/actionset.go [hunk 8]
```diff
@@ -342,6 +355,27 @@ func parseProfile(cmd *cobra.Command, ns string) (*crv1alpha1.ObjectReference, e
...
 
+// parseRepositoryServer returns the object reference
+// for the passed Repository Server
+func parseRepositoryServer(cmd *cobra.Command, ns string) (*crv1alpha1.ObjectReference, error) {
+	repositoryServerName, _ := cmd.Flags().GetString(repositoryServerFlagName)
+	if repositoryServerName == "" {
+		return nil, nil
+	}
+	if strings.Contains(repositoryServerName, "/") {
+		temp := strings.Split(repositoryServerName, "/")
```
Comment (viveksinghggits) [inline pkg/kanctl/actionset.go:366]:
```suggestion
		nsName := strings.Split(repositoryServerName, "/")
```

File: pkg/kanctl/actionset.go [hunk 9]
```diff
@@ -92,6 +94,7 @@ func newActionSetCmd() *cobra.Command {
 	cmd.Flags().StringSliceP(deploymentFlagName, "d", []string{}, "deployment for the action set, comma separated namespace/name pairs (eg: --deployment namespace1/name1,namespace2/name2)")
 	cmd.Flags().StringSliceP(optionsFlagName, "o", []string{}, "specify options for the action set, comma separated key=value pairs (eg: --options key1=value1,key2=value2)")
 	cmd.Flags().StringP(profileFlagName, "p", "", "profile for the action set")
+	cmd.Flags().StringP(repositoryServerFlagName, "r", "", "kopia repository server custom resource reference (eg: --repository-server namespace/name)")
```
Comment (viveksinghggits) [inline pkg/kanctl/actionset.go:97]: After looking at the rest of the code it looks like `--repository-server name` is also allowed. In that case do we want  expect the repo server to be available in the controller namespace or in the app namespace. 
Also, instead of supporting both the formats we should just support `namespace/name` format, just to avoid confusion.

File: pkg/kanctl/actionset.go [hunk 10]
```diff
@@ -727,3 +748,47 @@ func generateActionSetName(p *PerformParams) (string, error) {
 
 	return "", errMissingFieldActionName
 }
+
+func verifyRepositoryServerParams(repoServer *crv1alpha1.ObjectReference, crCli versioned.Interface, ctx context.Context) error {
+	if repoServer != nil {
+		rs, err := crCli.CrV1alpha1().RepositoryServers(repoServer.Namespace).Get(ctx, repoServer.Name, metav1.GetOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "Please make sure '%s' with name '%s' exists in namespace '%s'", "repository-server", repoServer.Name, repoServer.Namespace)
+		}
```
Comment (viveksinghggits) [inline pkg/kanctl/actionset.go:757]:
```suggestion
		if err != nil {
		 if apierrors.IsNotFound(err){
			return errors.Wrapf(err, "Please make sure '%s' with name '%s' exists in namespace '%s'", "repository-server", repoServer.Name, repoServer.Namespace)
			}
			return "there was an error getting repo server"
		}
```

**Review by viveksinghggits — COMMENTED** [review]
Almost, there.

File: pkg/kanctl/actionset.go (hunk 9 above)
Comment (r4rajat) [reply pkg/kanctl/actionset.go:98]: Yes @viveksinghggits, it is allowed but we want the repository server CR in controller namespace only

File: pkg/kanctl/actionset.go (hunk 6 above)
Comment (r4rajat) [reply pkg/kanctl/actionset.go]: Implemented !

File: pkg/kanctl/actionset.go [hunk 11]
```diff
@@ -342,6 +356,28 @@ func parseProfile(cmd *cobra.Command, ns string) (*crv1alpha1.ObjectReference, e
 	}, nil
 }
 
+// parseRepositoryServer returns the object reference
+// for the passed Repository Server
+func parseRepositoryServer(cmd *cobra.Command) (*crv1alpha1.ObjectReference, error) {
+	repositoryServerName, _ := cmd.Flags().GetString(repositoryServerFlagName)
+	if repositoryServerName == "" {
+		return nil, nil
```
Comment (akankshakumari393) [inline pkg/kanctl/actionset.go:364]: should we return an error here or an error message describing the issue instead of returning nil?

File: pkg/kanctl/actionset.go [hunk 12]
```diff
@@ -727,3 +750,50 @@ func generateActionSetName(p *PerformParams) (string, error) {
...
 }
+
+func verifyRepositoryServerParams(repoServer *crv1alpha1.ObjectReference, crCli versioned.Interface, ctx context.Context) error {
+	if repoServer != nil {
+		rs, err := crCli.CrV1alpha1().RepositoryServers(repoServer.Namespace).Get(ctx, repoServer.Name, metav1.GetOptions{})
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return errors.Wrapf(err, "Please make sure '%s' with name '%s' exists in namespace '%s'", "repository-server", repoServer.Name, repoServer.Namespace)
+			}
+			return errors.New("there was an error getting repo server")
```
Comment (akankshakumari393) [inline pkg/kanctl/actionset.go:761]:
```suggestion
			return errors.New("error while fetching repo server")
```

File: pkg/kanctl/actionset.go (hunk 11 above)
Comment (r4rajat) [reply pkg/kanctl/actionset.go:364]: We're handling errors where we're calling this function. I think we're doing similar while fetching other CRs. Let's see what others say.

File: pkg/param/param.go [hunk 13]
```diff
@@ -135,6 +137,21 @@ type KopiaServerCreds struct {
 	ConnectOptions map[string]int
 }
 
+// RepositoryServer contains where to store Repository Server CR Artifacts
```
Comment (viveksinghggits) [inline pkg/param/param.go:140]: This isn't necessarily true right?
We are going to store the repo server CRs in the etcd itself right?

File: pkg/kanctl/actionset.go (hunk 11 above)
Comment (viveksinghggits) [reply pkg/kanctl/actionset.go:364]: It's not about handling errors, even if we are handling errors where we are calling this function from; that would be of no use because we are not returning error.
Again, handling errors comes into picture when we return error.
One thing that I can think of is, we don't necessarily want to fail the `kanctl` command when repo server is not specified because there are chances that profile is specified. 
Can you please point me to the code where this check is performed to make sure either profile or repo server is  specified.

**Review by viveksinghggits — CHANGES_REQUESTED** [review]

File: pkg/param/param.go (hunk 13 above)
Comment (r4rajat) [reply pkg/param/param.go:140]: Yeah, so this RepositoryServer struct basically contains the selective artifacts from the Repo Server CR. It is not the Respo Server CR. It's only templating parameters.

File: pkg/kanctl/actionset.go (hunk 11 above)
Comment (r4rajat) [reply pkg/kanctl/actionset.go:364]: @viveksinghggits, While creating actionsets, we don't check/verify that either `--profile` or `--repository-server` flag is passed. in case either is passed, only then, it would verify the resources.

Comment (r4rajat) [reply pkg/kanctl/actionset.go:368]: Added check, if `--profile` or `--repository-server` flag is not provided, `kanctl` will warn user but will create actionset

File: pkg/param/param.go (hunk 13 above)
Comment (viveksinghggits) [reply pkg/param/param.go:140]: what are selective artifacts?

Comment (viveksinghggits) [inline pkg/param/param.go:140]:
```suggestion
// RepositoryServer contains fields from Repository server CR that will be used to resolve go templates for repository server in blueprint
```

**Review by viveksinghggits — APPROVED** [review]
